// db/actionDB.js
/**
 * Created by a.amrastanov on 21.12.2014.
 */

var util = require('util');
var TableDB = require('./tableDB');

function ActionDB(tableName){
    this.setTableName(tableName);
}
util.inherits(ActionDB, TableDB);


ActionDB.prototype.insertNew = function(relatedId, action, additional){
    var data = {
        relatedId: relatedId,
        actionTime: new Date(),
        action: action
    };
    if (additional !== undefined){
        data.additional = additional;
    }
    this.executeInsert(data);
};

ActionDB.prototype.insertNewWithTime = function(relatedId, action, actionTime, additional){
    var data = {
        relatedId: relatedId,
        actionTime: actionTime,
        action: action
    };
    if (additional !== undefined){
        data.additional = additional;
    }
    this.executeInsert(data);
};

var issueActionDb = new ActionDB('issue_action');
var orderActionDb = new ActionDB('order_action');

var ISSUE_ACTIONS = {
    HAND_CANCEL: 'hand_cancel',
    DISCONNECT_CANCEL: 'disconnect_cancel',
    END_BY_ORDER: 'end_by_order',
    GARBAGE_DELETE: 'garbage_delete'
};

var ORDER_ACTIONS = {
    TAKSIST_ACCEPT: 'taksist_accept',
    TAKSIST_REFUSE: 'taksist_refuse',
    CLIENT_REFUSE: 'client_refuse',
    ARRIVE_TO_CLIENT: 'arrive_to_client',
    CLIENT_DONT_COME: 'client_dont_come',
    GO_TO_TAKSIST: 'go_to_taksist',
    TRIP_START: 'trip_start',
    TRIP_END: 'trip_end'
};

module.exports.ISSUE_ACTIONS = ISSUE_ACTIONS;
module.exports.ORDER_ACTIONS = ORDER_ACTIONS;
module.exports.issueActionDb = issueActionDb;
module.exports.orderActionDb = orderActionDb;
