// db/dbHandler.js
/**
 * Created by a.amrastanov on 11.10.2014.
 */

var mysql = require('mysql');

var dbPool =  mysql.createPool({
        connectionLimit: 20,
        host: 'localhost',
        user: 'root',
        password: 'D00BwtUJWZRjLrjBuDb8feZyWIazSn',
        database: 'teztaksi',
        charset: 'utf8_general_ci'
});

module.exports.execute = function (queryString, fields, cb) {
    dbPool.getConnection(function (err, connection) {
        if (err) {
            if (cb){
                cb(require('errorCodes').get_db_connection_error);
            }
            require('out').error(err, queryString, fields);
        }
        else {
            connection.query(queryString, fields, function (err, result) {
                if (err) {
                    if (cb){
                        cb(require('errorCodes').db_query_error);
                    }
                    require('out').error(err, queryString, fields);
                }
                else if (cb){
                    cb(null, result);
                }
                connection.release();
            });
        }
    });
};

module.exports.executeInsert = function(tableName, fields, cb){
    var queryString = 'INSERT INTO '+ tableName + ' SET ?';
    module.exports.execute(queryString, fields, cb);
};

module.exports.executeUpdate = function(tableName, fields, whereFields, cb){
    var queryString = 'UPDATE '+ tableName + ' SET ? ';
    if (whereFields !== null){
        queryString += getEscapedWhereString(whereFields);
    }
    module.exports.execute(queryString, fields, cb);
};

module.exports.executeSelect = function(tableName, whereFields, cb){
    var queryString = 'SELECT * FROM '+ tableName;
    if (whereFields !== null){
        queryString += getEscapedWhereString(whereFields);
    }
    module.exports.execute(queryString, null, cb);
};

function getEscapedWhereString(whereFields){
    var query = "";
    for(var key in whereFields){
        query += " AND " + key + "=" + dbPool.escape(whereFields[key]);
    }
    if (query.length > 5){
        query = query.substr(5);
    }
    return " WHERE " + query + " ";
}
