// db/tableDB.js
/**
 * Created by a.amrastanov on 28.12.2014.
 */

function TableDB(tableName){
    if (tableName){
        this.setTableName(tableName);
    }
}

TableDB.prototype.dbHandler = require('./dbHandler');

TableDB.prototype.setTableName = function(tableName){
    this.tableName = tableName;
};

TableDB.prototype.getTableName = function(){
    return this.tableName;
};

TableDB.prototype.executeInsert = function(fields, cb){
    this.dbHandler.executeInsert(this.getTableName(), fields, cb);
};

TableDB.prototype.executeUpdate = function(fields, whereFields, cb){
  this.dbHandler.executeUpdate(this.getTableName(), fields, whereFields, cb);
};

TableDB.prototype.executeSelect = function(whereFields, cb){
    this.dbHandler.executeSelect(this.getTableName(), whereFields, cb);
};

TableDB.prototype.getFirst = function(whereFields, cb){
    this.executeSelect(whereFields, function(err, rows){
        if (err){
            cb(err);
        }
        else{
            if (rows === undefined || rows.length === 0){
                cb(require('errorCodes').no_field_in_db);
            }
            else{
                cb(null, rows[0])
            }
        }
    })
};


TableDB.prototype.getById =  function(id, cb){
    var whereFields = {id: id};
    this.getFirst(whereFields, cb);
};

TableDB.prototype.updateFieldById = function(id, updateField, cb){
    this.executeUpdate(updateField, {id: id}, cb);
};

module.exports = TableDB;
