// actionHandler/serverActions/sendTaksistArriveToClient.js
/**
 * Created by a.amrastanov on 24.09.2014.
 */

module.exports = function(parameterMap){
    this.action = parameterMap.action;
    this.order = parameterMap.order;
    this.taksist = parameterMap.taksist;
    this.doAction = function(){
        var order = this.order;
        var client = order.getClient();
        var actionHandler = require('actionHandler');
        var taksist = this.taksist;
        if (client){
            client.updateState(require('userHandler').WORK_STATE.TAKSIST_WAIT_CLIENT);
            var message = require('contextHandler').NewMessage(this.action);
            client.send(message);
            var sock = client.sock;
            sock.waitPing = true;
            setTimeout(function(){
                if (sock.waitPing){
                    delete sock.waitPing;
                    require('services').smsService.send(client, taksist);
                }
            }, 2000);
        }
        else{
            require('db').clientDB.getById(order.data.clientId, function (err, userData) {
                if (err){
                    require('out').error('error on get user data', err, 'tripStart', order);
                }
                else{
                    var UserConstructor = require('userHandler').clientHandler.userConstructor;
                    var user = new UserConstructor(userData);
                    require('services').smsService.send(user, taksist);
                    user.updateState(require('userHandler').WORK_STATE.TAKSIST_WAIT_CLIENT);
                }
            });
        }
    }
};
