// actionHandler/serverActions/sendOrderToTaksist.js
/**
 * Created by a.amrastanov on 06.08.2014.
 */

module.exports = function(parameterMap){
    this.order = parameterMap.order;
    this.taksist = parameterMap.taksist;
    this.action = parameterMap.action;
    this.doAction = function(){
        var issue = require('issueHandler').getIssueById(this.order.data.issueId);
        var orderToTaksistData = {
            id: this.order.data.id,
            issueId: issue.data.id,
            phone: this.order.getClient().data.phone
        };
        var message = require('contextHandler').NewMessage(this.action, {order: orderToTaksistData});
        this.taksist.send(message);
        var sock = this.taksist.sock;
        sock.waitPing = true;
        var order = this.order;
        var taksist = this.taksist;
        setTimeout(function(){
            if (sock.waitPing){
                delete sock.waitPing;
                var actionHandler = require('actionHandler');
                actionHandler.serverActionHandler.doAction({
                    action:actionHandler.serverActionsValues.sendOrderRefuseToClient,
                    order: order,
                    reason: actionHandler.serverActionConstants.NOT_AVAILABLE
                });
                taksist.removeOrder();
                require('orderHandler').removeOrder(order);
            }
        }, 10000);
    }
};
