// actionHandler/taksistActions/arriveToClient.js
/**
 * Created by a.amrastanov on 24.09.2014.
 */

module.exports = function(parameterMap){
    this.taksistToken = parameterMap.token;
    this.orderId = parameterMap.id;
    this.timestamp = parameterMap.timestamp;
    this.doAction = function(){
        var taksist = require('userHandler').taksistHandler.getUserByToken(this.taksistToken);
        taksist.saveCoordinate();
        taksist.updateState(require('userHandler').WORK_STATE.TAKSIST_WAIT_CLIENT);
        var order = require('orderHandler').getOrderById(this.orderId);
        if (order){
            var actionHandler = require('actionHandler');
            var actionTime = new Date(this.timestamp);
            actionHandler.serverActionHandler.doAction({
                action:actionHandler.serverActionsValues.sendTaksistArriveToClient,
                order: order,
                taksist: taksist
            });
            require('db').orderActionDb.insertNewWithTime(order.data.id, require('db').ORDER_ACTIONS.ARRIVE_TO_CLIENT, actionTime);
        }
    }
};
