// actionHandler/taksistActions/orderRefuse.js
/**
 * Created by Администратор on 05.07.2014.
 */

module.exports = function(parameterMap){
    this.taksistToken = parameterMap.token;
    this.orderId = parameterMap.id;
    this.reason = parameterMap.reason;
    this.timestamp = parameterMap.timestamp;
    this.doAction = function(){
        var taksist = require('userHandler').taksistHandler.getUserByToken(this.taksistToken);
        taksist.saveCoordinate();
        var orderHandler = require('orderHandler');
        var order = orderHandler.getOrderById(this.orderId);
        if (order){
            var actionHandler = require('actionHandler');
            var actionTime = new Date(this.timestamp);
            actionHandler.serverActionHandler.doAction({
                action:actionHandler.serverActionsValues.sendOrderRefuseToClient,
                order: order,
                reason: actionHandler.serverActionConstants.REFUSE
            });
            var client = order.getClient();
            if (client){
                actionHandler.serverActionHandler.doAction({
                    action:actionHandler.serverActionsValues.sendTaksistRemoveToClient,
                    taksist: taksist,
                    clients: [client]
                });
                var issue = client.getIssue();
                if(issue){
                    actionHandler.serverActionHandler.doAction({
                        action:actionHandler.serverActionsValues.sendIssueRemoveToTaksists,
                        issue: issue,
                        taksist: taksist
                    });
                    issue.removeTaksist(taksist);
                }
            }
            require('db').orderActionDb.insertNewWithTime(order.data.id, require('db').ORDER_ACTIONS.TAKSIST_REFUSE, actionTime, this.reason);
            taksist.removeOrder();
            orderHandler.removeOrder(order);
        }
    }
};
