// actionHandler/taksistActions/setTaksistCoordinate.js
/**
 * Created by a.amrastanov on 05.07.2014.
 */

module.exports = function(parameterMap){
    this.taksistToken = parameterMap.token;
    this.lat = parameterMap.lat;
    this.lon = parameterMap.lon;
    this.doAction = function(){
        var userHandler = require('userHandler');
        var taksistHandler = userHandler.taksistHandler;
        var taksist = taksistHandler.getUserByToken(this.taksistToken);
        taksist.setCoordinate(this.lat, this.lon);
        var clients = taksist.getAllCoordinateRequestedClients();
        var taksistShowRequestedUsers = userHandler.clientHandler.getTaksistsShowRequestedUsers();
        clients = clients.concat(taksistShowRequestedUsers);
        var order = taksist.getOrder();
        if (order !== undefined){
            var orderClient = order.getClient();
            clients = clients.concat(orderClient);
        }
        if (clients.length > 0){
            var actionHandler = require('actionHandler');
            actionHandler.serverActionHandler.doAction({
                action:actionHandler.serverActionsValues.sendTaksistCoordinateToClient,
                taksists: [taksist],
                clients: clients
            });
        }
    }
};
