// actionHandler/taksistActions/tripEnd.js
/**
 * Created by a.amrastanov on 29.09.2014.
 */

module.exports = function(parameterMap){
    this.taksistToken = parameterMap.token;
    this.orderId = parameterMap.id;
    this.distance = parameterMap.distance;
    this.timestamp = parameterMap.timestamp;
    this.doAction = function(){
        var taksist = require('userHandler').taksistHandler.getUserByToken(this.taksistToken);
        taksist.saveCoordinate();
        var order = require('orderHandler').getOrderById(this.orderId);
        if (order){
            order.setDistance(this.distance);
            order.saveDistance();
            var actionTime = new Date(this.timestamp);
            var actionHandler = require('actionHandler');
            actionHandler.serverActionHandler.doAction({
                action:actionHandler.serverActionsValues.sendTripEndDetailsToClient,
                order: order
            });
            require('db').orderActionDb.insertNewWithTime(order.data.id, require('db').ORDER_ACTIONS.TRIP_END, actionTime);
            taksist.removeOrder();
            require('orderHandler').removeOrder(order);
        }
    }
};
