// actionHandler/clientActions/createIssue.js
/**
 * Created by Администратор on 05.07.2014.
 */

module.exports = function(parameterMap){
    this.clientToken = parameterMap.token;
    this.region = parameterMap.region;
    this.addressFrom = parameterMap.addressFrom;
    this.addressTo = parameterMap.addressTo;
    this.orientation = parameterMap.orientation;
    this.lat = parameterMap.lat;
    this.lon = parameterMap.lon;
    this.offeredPrice = parameterMap.offeredPrice;
    this.doAction = function(){
        var userHandler = require('userHandler');
        var client = userHandler.clientHandler.getUserByToken(this.clientToken);
        var oldIssue = client.getIssue();
        var actionHandler = require('actionHandler');
        if (oldIssue){
            actionHandler.serverActionHandler.doAction({
                action:actionHandler.serverActionsValues.sendIssueRemoveToTaksists,
                issue: oldIssue
            });
            oldIssue.removeUsers();
            require('db').issueActionDb.insertNew(oldIssue.data.id, require('db').ISSUE_ACTIONS.HAND_CANCEL);
            require('issueHandler').removeIssue(oldIssue);
        }
        require('issueHandler').generateIssue(client.data.id, this.region, this.addressFrom, this.addressTo, this.lat, this.lon, this.orientation, this.offeredPrice);
    }
};
